### ISSUE TITLE
Permalinks

### ISSUE THREAD

**Issue by tmcw** [openlawdc/browser#1]

**Comment by tmcw** [comment]
straw-man urls:

```
http://dccode.org/browse/#10-10.2
```

With a custom 404 handler or with tons of HTML generation it should be possible to support this:

```
http://dccode.org/browse/10-10.2
```

Already we have basically

```
http://dccode.org/section/10-10.2.json
```

And we should have

```
http://dccode.org/section/10-10.2.txt
http://dccode.org/section/10-10.2.html
```

**Comment by konklone** [comment]
I'm sure the hashes are easier to support given the site's static architecture, though I'm always antsy about supporting hash-driven URLs, with or without the bang, indefinitely into the future. It places more limits on how the site's tech can evolve.

**Comment by waldoj** [comment]
FWIW, I've found a very popular feature in The State Decoded is this style of URL:

http://dccode.org/10-10.2/

This enables folks to use the URL as a command line—just enter the section number that you want after the domain name. Attorneys just love it.

**Comment by tmcw** [comment]
y, I threw in a /browse/ there because I don't want to crowd `/` prematurely, but maybe if we always know the id format it should be fine.

**Comment by waldoj** [comment]
Yeah, I wrestled with the decision of the URL format for weeks. I'm happy with my decision, but I'm not _comfortable_ with it. :)

**Comment by tmcw** [comment]
Okay, adding:

I want this to refer to a version of the code, not just 'a section and id' which has content that can change.

```
http://dccode.org/browse/0/10-10.2
```

Let's talk sunday about what versions can be or are.

Another nice-to-have is deeper linking, to paragraphs, sentences, and so on. One example is [what Dave Winer is doing](http://scripting.com/stories/2010/06/23/improvingParagraphlevelPer.html#inTheNextIterationTheNamesWereABitNicerAsItWasRenderingTheStoryItWouldBumpACounterEachTimeItMovedToANewParagraphSoTheNamesWouldBeP1P2EtcTheOnlyProblemIsIfIAddedAParagraphInTheMiddleAllTheNumbersWouldMoveDownAndLinksMightPointToTheWrongPlaceWhichDefeatsThePurposeOfHavingThePermalinks), but his is tuned so that posts can be updated. If we have non-updateable versions, then we can use something much simpler, like `charstart-charend`.

**Comment by konklone** [comment]
Interestingly, it looks like Dave Winer's changed his permalinks from what's described in that post - his paragraph-level permalinks now appear to be the actual paragraph text, condensed. Which wouldn't work if a paragraph was updated, unless the anchor is coded to stay static after initial publication, or something. You could still use the approach he wrote about, of course. :)

Having a version in the URL would make paragraphs immutable, so you have a lot more flexibility in anchoring them. But only if you don't use default or dynamic versions, like "current".

**Comment by waldoj** [comment]
There are two ways to approach the storage of older versions of laws. The first is to figure that basically nobody cares about what laws used to say—they only care about what it is _now_—and that there's simply no point in storing older versions of laws. Better to link to bills from the histories. The second is to figure that it's useful information, worth storing, and the diffs could be valuable.

Once one resolves to store laws, then there are two ways to approach the URL structure for them. The first is to figure that, since basically nobody cares about old versions, there's no point in polluting the URL structure with them. That is, using `http://example.com/2011/1.23-45/` instead of `http://example.com/1.23-45/` isn't actually helpful because people either don't want to read the law as it existed two years ago or, worse, they won't understand that they're looking at an old version of the law that no longer applies. In that instance, best to provide versions in the form of URL fragments, e.g. `http://example.com/1.23-45/#revision-2011`. The second approach is to figure that the years are so wildly important that they need to be stored within the URLs, and that people will use them as a primary method of URL navigation, so that they can see what laws used to look like.

For The State Decoded, I support storage of older versions of laws, although there's no support for displaying them in the UI yet. I haven't determined what the URL structure will look like, but it will not involve inserting the year of the revision to the left of the section number, because a) it would confuse people and b) basically nobody cares.

**Comment by waldoj** [comment]
For deeper linking, remember that this is precisely why there are subsection identifiers—`2(a)6` and that sort of thing. Lawmakers long ago realized the need for this, devised a standardized method of accomplishing this, and then used that as the common identifier for the most fundamental units of laws (subsections) that are cited in laws, court decisions, and legislation. In The State Decoded, those are used directly in the URL fragments. Want to link to `§ 123-4.5(a)B1`? That's at `http://example.com/123-4.5#(a)B1`.

**Comment by konklone** [comment]
One way to view this is: what will other people/services need to know to construct URLs to my service? If you require a version identifier in your URLs, then anyone who wants to automatically link to your DC Code browser has to have that year at link-time. My suspicion is that will be burdensome and tricky in the long run for a lot of uses.

**Comment by waldoj** [comment]
That is an awfully interesting point that never crossed my mind.

**Comment by tmcw** [comment]
> One way to view this is: what will other people/services need to know to construct URLs to my service?

Sure, the 'software creating links' use case is usually covered by a `/latest/` version, which can either be a mirror of the latest specific version or a redirect to it.

> isn't actually helpful because people either don't want to read the law as it existed two years ago or, worse, they won't understand that they're looking at an old version of the law that no longer applies

There's the 'viewing an older version' pattern - [here's an example](http://mapbox.com/mapbox.js/api/v0.6.6/).

**Comment by waldoj** [comment]
> There's the 'viewing an older version' pattern - here's an example.

That's only useful if the link is to the top of the page, rather than an anchor. Otherwise you need some kind of an overlay.

I've also come to find that, at least with the Florida and Virginia sites, the bulk of the people reading laws online are fairly uneducated. (That conclusion is based on search terms in the referrer logs, search terms used within Solr, comments, and e-mails.) The sort of person who is googling for legal information is likely facing civil or criminal charges (or, frequently, has a family member or significant other who is), and cannot afford an attorney. As I have come to realize this, it has affected pretty strongly many of [the UI](github.com/meticulous/state-decoded-ui-2012) decisions and features that I've planned and implemented. I simply ask myself if your average eighth grader could understand it. This isn't the audience that I've anticipated but, as I've talked more with folks from the legal aid world, it's now blindingly obvious, in retrospect. That's another reason why I've basically punted on displaying older versions of laws in the UI (as opposed to via the API or in bulk downloads)—it's just going to make it harder for these folks. I can put up all of the notices and disclaimers and so on that I want, but people are still not going to understand that what they're reading is not, in fact, the law.

That, of course, is my approach—it need not be yours!

**Comment by tmcw** [comment]
Good points & good background in general - will think about this and let's talk Sunday if you can make it!

**Comment by waldoj** [comment]
Oh, yeah—@krues8dr and I will be taking the 7 AM train up, and be there all day.

**Comment by konklone** [comment]
Looks like this is far enough along that this issue can be closed, and refinements taken up in newer specific threads.

**Comment by schmod** [comment]
The way that we're doing this right now really isn't ideal.  In the immediate future, I'd like to tie your citation parser into the URL router.   We need to do that before we can start working on a bunch of other open bugs.  

After that, my vote would be to format URLs in the form of /[citation]/[version], and default to the most recent version it's not explicitly specified.

**Comment by konklone** [comment]
Isn't my citation extractor a bit overkill to parse the URL? Seems like we could keep it in a nice, predictable format.

There's discussion above in this thread about having a version in the URL, but having it show up at the beginning, like /current/[citation], or /1981/[citation], etc. Hierarchically, that maps a little more cleanly to what the URL is referencing.

**Comment by schmod** [comment]
I agree that it makes more sense hierarchically, but it seems less convenient for what is (presumably) the most common use case, where you'd want to link to the current version of the code.

**Comment by konklone** [comment]
I think it'd be more or less the same if the version defaulted to `current` or `latest`, with the exception of someone typing in the URL manually. I don't have strong feelings on it. But we might want to have things after the main section, such as anchor tags to the credits, history, and individual paragraphs. With the version on the end, you might end up with `/3-102/latest#(a)`, and so the version is splitting the citation.

**Comment by krusynth** [comment]
With most APIs, version usually comes before the resource-specifying parameters.  Except with facebook's API, notably.  Which isn't to say that it _has_ to be done that way, but semantically it makes more sense, I think?  /data/current/this-thing - you want a particular item from a collection that spans many things from a particular time period, rather than, I want a particular version of a thing?  

https://blog.apigee.com/detail/restful_api_design

OTOH, it's also safe to assume that laws change individually, so each law has its own revisions?  That doesn't match the backend structures we've created over on SD, but is definitely a valid use case.

**Comment by konklone** [comment]
I don't think the laws change individually? @vzvenyach might know more, but I believe there's the "working version" of the Code, and then a few major historical editions (I see the "1981" edition a lot in section credits). But I don't think individual sections would have an independent versioning system that would need to be reflected in the URL.
